### DISCUSSION TITLE
Events from custom nodes

### DISCUSSION THREAD

**Discussion by SciManJD** [bcakmakoglu/vue-flow#886] [Q&A] [answered]
I have some problems to understand the complete concept concerning the handling of the events from custom nodes. It is clear how to send an event from the node. But looking at the documentation and examples, I couldn't understand how to catch them. If I have several instances of one of my custom nodes, how can I find out which one sends the event? I think that probably I can't find a way to send also the node ID, but I couldn't find such example. Can I get a short example?

**Comment by bcakmakoglu** [comment]
Mh? Which event are you talking about specifically?

In the end it's really just something like this
```ts
onNodeClick(({ node, event, connectedEdges }) => {
	console.log(node)

	if (node.id === id) {
		// do something
	}
})

**Reply by SciManJD** [reply]
As I said, I have some problems with the understanding, therefor I'm looking for an example. I have a radio buttons in the custom node. I emit an event when the state of the selection is changed with emit(). I want to get the state of these radio buttons. I there is only one instance of the node, I can get the event that is sent by emitter.on(). But if there are several instances, I don't know which one sends the event.

**Reply by bcakmakoglu** [reply]
Well, send the id with your event to differentiate which node it came from then, this is really up to you and not really related to the library 😄 

For example

```ts
const onInputChange = (value) => {
	emit('input', { value, id })
}
```

```vue
<script setup>
const handleInputChange = ({ value, id }) => {
	// do something
}
</script>

<VueFlow v-model="elements">
	<template #node-custom="props">
		<CustomNode v-bind="props" @input="handleInputChange" />
	</template>
</VueFlow>
```

**Reply by SciManJD** [reply]
I have some more stupid questions. Initially I used the emitter of Vue. But is seems that the component has own function emit. So I made the following (sorry, I said radio buttons, but it is a checkbox actually):

in the new node:

```
<template>
   ...
      <input type="checkbox" v-model="state" @change="stateChanged()">
   ...
</template>

<script>
   ...
   methods: {
       stateChanged() {
                this.$emit('polarity', {state: this.state, id: 2});
       }
   }
   ...
</script>
```

And in the parent component:

```
<template>
   <vue-flow @nodeClick="nodeSelected($event)">
      <template #node-custom>
         <vue-flow-custom @input="handInputChange"/>
      </template>
   </vue-flow>
</template>

<script>
   methods: {
      handleInputChange(event) {
         console.log(JSON.stringify(event));
      }
   }
</script>

```
When I change the state of the checkbox, I receive the event, but it is always {"isTrusted":true}, instead of the state and ID.
And another question... Where it comes this ID, when emit the event? Show I get it with a function?

**Reply by SciManJD** [reply]
I found the problem. I send event with name "polarity", but expect an event "input". Now it works fine. But I'm still wondering where it comes the ID.

**Reply by SciManJD** [reply]
I solved the problem with the ID in the following way. I send the ID via the props, then I return it with the emit(). It works fine, but is this the normal way?

**Reply by bcakmakoglu** [reply, marked as answer]
It's the intended way, yeah

**Comment by SciManJD** [comment]
Hi again! I still cannot understand something. As I wrote above, now I can send and receive the event. But I always get two events - one with content {"isTrusted":true,"_vts":1683375379380} and after it follows my event. I read about this 'isTrusted' property, but it should be a property of my event. In my case it is in another separated event. So I suspect that I'm doing something not completely correct. But what it could be?

**Reply by bcakmakoglu** [reply]
Can you give me a snippet of your code?
What event are you binding on your custom node? So what `@event` are you listening for?

**Comment by SciManJD** [comment]
The code is already given above. I have a checkbox in my custom node. It seems that this first event is send when the checkbox state is changed. The second event is sent by my emit. But I couldn't find information how can I disable this event, generated by the checkbox.
